src/producer.rs

```diff
@@ -216,26 +179,57 @@ fn zip_producer(tmp_dir: &Path, zip_files: &[&String], queue: &WorkQueue, ignore
 }
 
 pub fn producer(tmp_dir: &Path, paths: &[String], queue: &WorkQueue, ignore_orphan_gcno: bool, is_llvm: bool) -> Option<Vec<u8>> {
-    let mut zip_files = Vec::new();
     let mut directories = Vec::new();
 
+    let mut gcno_archive: Option<ZipArchive<File>> = None;
+    let mut gcda_archives: Vec<(&String, Archive)> = Vec::new();
+    let mut info_archives: Vec<(&String, ZipArchive<File>)> = Vec::new();
+
     for path in paths {
         if path.ends_with(".zip") {
-            zip_files.push(path);
+            let archive = open_archive(path);
+            if path.contains("gcno") {
+                gcno_archive = Some(archive);
+            } else if path.contains("gcda") {
+                gcda_archives.push((path, Archive {
+                    item: ArchiveType::Zip(archive),
+                }));
+            } else if path.contains("info") || path.contains("grcov") || path.contains("jsvm") {
+                info_archives.push((path, archive));
+            } else {
+                panic!("Unsupported archive type.");
+            }
         } else {
             directories.push(path);
         }
     }
 
-    let ret1 = zip_producer(tmp_dir, &zip_files, queue, ignore_orphan_gcno, is_llvm);
-    let ret2 = dir_producer(&directories, queue, ignore_orphan_gcno);
-
-    if ret1.is_some() {
-        ret1
-    } else if ret2.is_some() {
-        ret2
+    if gcno_archive.is_some() && gcda_archives.is_empty() && !directories.is_empty() {
```

Comment:
Do you think we could have a single producer that loops in all archives (where an "archive" would be either a Zip or a Directory) and generates data from all archives, instead of splitting it into two (zip_producer and dir_producer)?